subsys/bluetooth/audio/has.c

```diff
@@ -1200,6 +1289,76 @@ int bt_has_register(const struct bt_has_register_param *param)
 		has.features |= BT_HAS_FEAT_WRITABLE_PRESETS_SUPP;
 	}
 
+	return 0;
+}
+
+int bt_has_features_set(const struct bt_has_features_param *features)
+{
+	uint8_t tmp_features;
+	int err;
+
+	/* Check whether any features will change, otherwise we don't want to notify clients */
+	if (has.features != features->type ||
+	    (has.features & BT_HAS_FEAT_PRESET_SYNC_SUPP) ^ features->preset_sync_support ||
+	    (has.features & BT_HAS_FEAT_INDEPENDENT_PRESETS) ^ features->independent_presets) {
+		return 0;
+	}
+
+	tmp_features = has.features;
+
+	err = has_features_register(features);
+	if (err != 0) {
+		LOG_DBG("Failed to register features");
+		return err;
+	}
+
+#if defined(CONFIG_BT_HAS_FEATURES_NOTIFIABLE)
+	bool tmp_pending_ntf_features[ARRAY_SIZE(has_client_list)];
+
+	for (size_t i = 0U; i < ARRAY_SIZE(has_client_list); i++) {
+		struct has_client *client = &has_client_list[i];
+		/* save old state */
+		tmp_pending_ntf_features[i] = client->pending_ntf.features;
+		/* mark to notify */
+		client->pending_ntf.features = true;
+	}
+
+	err = k_work_submit(&features_work);
+	if (err != 0) {
+		/* restore old values */
+		for (size_t i = 0U; i < ARRAY_SIZE(has_client_list); i++) {
+			struct has_client *client = &has_client_list[i];
+			client->pending_ntf.features = tmp_pending_ntf_features[i];
+		}
+		has.features = tmp_features;
+	}
+#endif /* CONFIG_BT_HAS_FEATURES_NOTIFIABLE */
+
+	return err;
+}
```

Comment:
I wonder whether this whole function should be guarded with `CONFIG_BT_HAS_FEATURES_NOTIFIABLE`.
Meaning that if features can be changed then the characteristic is notified. On the other hand, if the features remain unchanged (meaning the API does not allow to change it), the characteristic is not notifiable.

Reply:
I get your point, however, from a strict spec perspective it is allowed to change the features without notifying. Based on that we should expose the API function regardless. I don't think it hurts that it is there - and I don't think we are optimizing anything by removing it either.

Reply:
It's more about functionality. Imagine, the `Preset Synchronization Support` will be disabled in at some point. The client shall then write to Control Point of both Hearing Aids the new preset.
I think that if the features can be changed, the client shall be notified about the change. I think that was the intention of having the Features characteristic optionally notifiable.

Reply:
It's a fair point